### PR TITLE
LP-2446 sync user data with mailchimp

### PR DESCRIPTION
**Pull request by asadali145**
[LP-2446](https://philanthropyu.atlassian.net/browse/LP-2446)

### CODE REVIEW

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/574/workflows/workflows_id/jobs/574/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tasks.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   139 lines
Missing: 0 lines
Coverage: 100%
-------------
```

File: openedx/adg/common/mailchimp_pipeline/client.py [hunk 1]
```diff
@@ -1,6 +1,4 @@
-"""
-Mailchimp client
-"""
+"""Mailchimp client"""
```
Comment (zahid-arbisoft) [inline openedx/adg/common/mailchimp_pipeline/client.py:1]: As per [edx style guide](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#docstrings) the preferable style is 

```
"""
Docstring here
"""
```

File: openedx/adg/common/mailchimp_pipeline/handlers.py [hunk 2]
```diff
@@ -1,33 +1,44 @@
-"""
-Signals for Mailchimp pipeline
-"""
+"""Signals for Mailchimp pipeline"""
```
Comment (zahid-arbisoft) [inline openedx/adg/common/mailchimp_pipeline/handlers.py:1]: same here

File: openedx/adg/common/mailchimp_pipeline/helpers.py [hunk 3]
```diff
@@ -22,4 +20,4 @@ def get_enrollment_course_names_and_short_ids_by_user(user):
         return '', ''
 
     short_ids, display_names = zip(*enrollments)
-    return ','.join(map(str, short_ids)), ','.join(display_names)
+    return ', '.join(map(str, short_ids)), ', '.join(display_names)
```
Comment (zahid-arbisoft) [inline openedx/adg/common/mailchimp_pipeline/helpers.py:23]: Lets revert this change, we explicitly removed the spaces

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/575/workflows/workflows_id/jobs/575/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tasks.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   139 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Review by zahid-arbisoft — CHANGES_REQUESTED** [review]

File: openedx/adg/common/mailchimp_pipeline/client.py (hunk 1 above)
Comment (asadali145) [reply openedx/adg/common/mailchimp_pipeline/client.py:1]: Do we prefer edx guidelines over pep8?

File: openedx/adg/common/mailchimp_pipeline/helpers.py (hunk 3 above)
Comment (asadali145) [reply openedx/adg/common/mailchimp_pipeline/helpers.py:23]: A test was failing due to this. I will revert this and will change the test. It was an inconsistency between test and this method.

Comment (zahid-arbisoft) [reply openedx/adg/common/mailchimp_pipeline/helpers.py:23]: ok.

File: openedx/adg/common/mailchimp_pipeline/client.py (hunk 1 above)
Comment (zahid-arbisoft) [reply openedx/adg/common/mailchimp_pipeline/client.py:1]: I think pep8 does not give error or warning over this

Comment (asadali145) [reply openedx/adg/common/mailchimp_pipeline/client.py:1]: Yes, Actually it is preferred to use single line in pep8 guidelines if docstring is a one liner, It is also inconsistent in open edX code :)

I am changing to the edX preferred style for this PR.

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/576/workflows/workflows_id/jobs/576/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tasks.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   143 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/577/workflows/workflows_id/jobs/577/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tasks.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   143 lines
Missing: 0 lines
Coverage: 100%
-------------
```

File: openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py [hunk 4]
```diff
@@ -36,5 +36,5 @@ def test_get_enrollment_course_names_and_short_ids_by_user(user_enrollments):  #
     user, _ = user_enrollments
     course_short_ids, course_titles = get_enrollment_course_names_and_short_ids_by_user(user)
 
-    assert course_short_ids == '100, 101, 102'
-    assert course_titles == 'course1, course2, course3'
+    assert course_short_ids == '100,101,102'
+    assert course_titles == 'course1,course2,course3'
```
Comment (ahmed-faraz46) [inline openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py:40]: It is fixed in @annas-usman's [PR](https://philanthropyu.atlassian.net/browse/LP-2458) as well, so we would have to revert from one file, and use we can manually check that this particular test is failing due to this issue and would be fixed

File: openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py [hunk 5]
```diff
@@ -2,34 +2,110 @@
...
+
+@pytest.mark.django_db
+@pytest.mark.mute_signals
+@skip_unless_lms
+def test_task_send_user_info_to_mailchimp_user_application_created(mocker):
+    """
+    Assert that mailchimp client is called with valid data when User is created.
+    """
+    mock_mailchimp_client = mocker.patch.object(mailchimp_tasks, 'MailchimpClient', autospec=True)
+    mock_create_or_update_list_member = mock_mailchimp_client().create_or_update_list_member
```
Comment (ahmed-faraz46) [inline openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py:72]: These two lines are repeated in all tests let's please define a fixture and use it in all these tests.

**Review by ahmed-faraz46 — CHANGES_REQUESTED** [review]

File: openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (hunk 4 above)
Comment (asadali145) [reply openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py:40]: Okay then I am reverting this one.

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/579/workflows/workflows_id/jobs/579/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tasks.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   105 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/580/workflows/workflows_id/jobs/580/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tasks.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   105 lines
Missing: 0 lines
Coverage: 100%
-------------
```

File: openedx/adg/common/mailchimp_pipeline/helpers.py [hunk 6]
```diff
@@ -23,3 +23,77 @@ def get_enrollment_course_names_and_short_ids_by_user(user):
...
+
+    Returns:
+        bool: True if sync is required else False.
+    """
+    MAILCHIMP_FIELDS = ['organization', 'status', 'business_line', 'company']
+    update_fields = kwargs.get('update_fields') or []
+    return created or sender == UserProfile or any(field in update_fields for field in MAILCHIMP_FIELDS)
+
+
+def get_user_merge_fields(instance):
```
Comment (ahmed-faraz46) [inline openedx/adg/common/mailchimp_pipeline/helpers.py:46]: Now as we are sure the instance which it would be receiving is a user instance. So we can give it a specific name.

Comment (ahmed-faraz46) [reply openedx/adg/common/mailchimp_pipeline/helpers.py:46]: Same goes for all these helpers.

File: openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py [hunk 7]
```diff
@@ -4,7 +4,17 @@
 import pytest
 from django.test.utils import override_settings
 
+from openedx.adg.common.mailchimp_pipeline.helpers import (
```
Comment (ahmed-faraz46) [inline openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py:7]: We should write separate tests for helpers as well.

**Review by ahmed-faraz46 — CHANGES_REQUESTED** [review]

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/581/workflows/workflows_id/jobs/581/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tasks.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_handlers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_tasks.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
openedx/adg/lms/registration_extension/tests/factories.py (100%)
-------------
Total:   145 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Review by ahmed-faraz46 — APPROVED** [review]
LGTM 👍
